### PR TITLE
QueryDispatcher/Request.js refactor

### PR DESCRIPTION
**Pull request by leonazava**
**In this PR:**
* Refactored QueryDispatcher/Request.js using async/await syntax
* Fixed error handling & formatting
* Made methods more semantically precise - methods only do what their method name suggests they are

### CODE REVIEW

**Comment by aleksandrakorolova** [comment]
1. Can reproduce issue from the task:
- URL for Men category
- Category Description for Tech category
- Products from Women category
Screenshot 2022-07-13 at 15 21 28

2.Error fetching messages don't appear **BUT** Error message 'Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'message')' appears in the console

File: packages/scandipwa/src/util/Request/QueryDispatcher.js [hunk 1]
```diff
@@ -57,7 +59,10 @@ export class QueryDispatcher {
             this.promise = await executeGet(prepareQuery(queries), name, cacheTTL, this.controller.signal);
             this.onSuccess(this.promise, dispatch, options);
         } catch (err) {
-            this.onError(err, dispatch, options);
+            if (err.message !== REQUEST_ABORT_MESSAGE
+                && err.message !== REQUEST_ABORT_MESSAGE_2) {
+                this.onError(err, dispatch, options);
+            }
```
Comment (zans-laksa) [inline packages/scandipwa/src/util/Request/QueryDispatcher.js:65]: Can we trust these strings? Can they change on different browsers and versions?

Comment (zans-laksa) [reply packages/scandipwa/src/util/Request/QueryDispatcher.js:65]: Maybe there is some error code available?

Comment (zans-laksa) [reply packages/scandipwa/src/util/Request/QueryDispatcher.js:65]: `REQUEST_ABORT_MESSAGE` as a substring of `REQUEST_ABORT_MESSAGE_2`.
So actually you don't need two constants, you can just search a substring.

**Review by zans-laksa — CHANGES_REQUESTED** [review]

**Review by zans-laksa — APPROVED** [review]

**Comment by aleksandrakorolova** [comment]
1. Still can reproduce different cases:

- URL and category description are displayed for category X but products are displayed for category Y
Screenshot 2022-07-18 at 15 55 55
- URL and products are displayed for category X but category description for category Y
Screenshot 2022-07-18 at 15 59 21
- URL is displayed for category X but category description and products for category Y
Screenshot 2022-07-18 at 16 08 52

2. Fixed

**Review by zans-laksa — APPROVED** [review]

**Comment by leonazava** [comment]
Relevant for #4974 and #3088:
Those issues are a byproduct of our caching strategy. 
I initially planned to cancel a request if a newer request was dispatched before the previous one completes. The problem is that the cache is... to fast. Getting a response from the cache only takes 2-3ms, which is way faster than a human is physically able to click on a new link to dispatch a new request and cancel an older one. 
This becomes problematic when the service worker, which sends requests to an actual server, on slow 3g takes anywhere from 200ms-6s to complete a request. A user can queue multiple uncancellable requests like that, and they all will resolve and attempt to update the app state, which leads to the behavior described in the aforementioned issues.

This PR won't solve the issues but it still has a lot of good code, reactors, adjustments, etc, so I believe we could still try to merge it.
How to test: this PR doesn't add anything new behavior-wise. As long as homepage/pdp/plp/cart/checkout work as intended, it's safe to merge.

**Comment by Ylmzef** [comment]
I'm still facing with this issue in 5.3.3

**Comment by sliceem88** [comment]
Getting error on search with this fix:
`code: "410"
error: true
message: "Query hash is unknown"`
image

Rolled back to 5.3.3

**Comment by leonazava** [comment]
@Ylmzef this PR won't fix the issues, refer to [the earlier comment](https://github.com/scandipwa/scandipwa/pull/5028#issuecomment-1189990491)

@sliceem88 Is the search result broken as well? If it's just the log, you're most probably seeing it because of better error handling and the cause of the error exists in the older versions too. There are also a few edge cases that cause the search to fail that are unrelated to this PR.

**Comment by Ylmzef** [comment]
> Getting error on search with this fix: `code: "410" error: true message: "Query hash is unknown"` image
> 
> Rolled back to 5.3.3

Same here

**Comment by sliceem88** [comment]
> @Ylmzef this PR won't fix the issues, refer to [the earlier comment](https://github.com/scandipwa/scandipwa/pull/5028#issuecomment-1189990491)
> 
> @sliceem88 Is the search result broken as well? If it's just the log, you're most probably seeing it because of better error handling and the cause of the error exists in the older versions too. There are also a few edge cases that cause the search to fail that are unrelated to this PR.

Any reason why rolled back worked ?

**Comment by Ylmzef** [comment]
Yes, I rolled back to 5.3.3 and it's working fine again.

**Comment by Ylmzef** [comment]
is there any progress yet? Our website not available to customers for a long time due this issue.
Maybe can you release a patch for this issue? It’s not a simple problem for live shops.

**Comment by leonazava** [comment]
@Ylmzef the fix can be found [here](https://github.com/scandipwa/scandipwa/issues/5181).

**Pull request merged**